### PR TITLE
Refactor how we handle overflow and remove blanket Pattern impl

### PR DESCRIPTION
**Pull request by nikomatsakis**
See comments in the commits for reasoning

r? @aturon 
cc @Kimundi

### CODE REVIEW

File: src/libcore/str/pattern.rs [hunk 1]
```diff
@@ -474,22 +474,16 @@ impl<'a, 'b> Pattern<'a> for &'b [char] {
                       s, CharEqPattern(s));
 }
 
+/// Searches for chars that are equal to any of the chars in the array
```
Comment (Kimundi) [inline src/libcore/str/pattern.rs]: This should probably be the doc comment of the `&str` impl instead.

File: src/libcollections/string.rs [hunk 2]
```diff
@@ -765,6 +766,25 @@ impl<'a> Extend<&'a str> for String {
     }
 }
 
+// Convenience method so that you can pass in a `&String` as a pattern.
```
Comment (Kimundi) [inline src/libcollections/string.rs]: This could also be a doc comment explaining that it delegates to the `&str` impl.

**Comment by Kimundi** [comment]
Apart from the doc comments, the pattern impls look fine to me.

**Comment by alexcrichton** [comment]
Is it possible to still have a forwarding impl, but just for references?

```rust
impl<'a, 'b, P: Pattern<'a> + ?Sized + 'b> Pattern<'a> for &'b P {
    ...
}
```

**Comment by nikomatsakis** [comment]
@alexcrichton no, because any type `P: Fn` is also `P: Pattern`, and if we make it so that `&F` is `Fn`, then you have two overlapping routes to derive that `&P: Pattern`.

**Comment by nikomatsakis** [comment]
rebased, tweaked docs per @Kimundi's suggestion

**Comment by freebroccolo** [comment]
I think the overflow refactoring here might be what @aturon was referring to in the issue I reported at #23714. The overflow/failure on evaluating the `Sized` bound causes my [ackermann.rs](https://github.com/freebroccolo/ackermann.rs) crate to no longer compile since I am using `Cow` in a rose-tree like structure [here](https://github.com/freebroccolo/ackermann.rs/blob/master/src/num.rs#L11-L20) which triggers the issue.

**Pull request merged**